lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
-                                                 UsedIterator->Location,
-                                                 TheBinary);
-          ReturnStruct->Fields.insert(std::move(CurrentField));
-
-          ReturnStruct->Size += model::Register::getSize(Register);
-        } else if (MustUseTheNextOne) {
-          if constexpr (!AT::OnlyStartDoubleArgumentsFromAnEvenRegister)
-            return std::nullopt;
-          else if ((Index & 1) == 0 || ReturnStruct->Fields.size() <= 1
-                   || Index <= 1)
-            return std::nullopt;
-        }
-
-        MustUseTheNextOne = MustUseTheNextOne || IsCurrentRegisterUsed;
+      for (const auto &MaybeType : ResultingTypes) {
+        model::QualifiedType Qualified;
+        if (MaybeType.has_value())
+          Qualified = MaybeType.value();
+
+        model::StructField CurrentField;
+        CurrentField.Offset = ReturnStruct->Size;
+        CurrentField.Type = ensureTypeIsAvailable(Qualified,
+                                                  Allowed.at(0),
+                                                  Binary,
+                                                  TypeDependencies);
+        ReturnStruct->Fields.insert(std::move(CurrentField));
+
+        auto MaybeSize = CurrentField.Type.size();
+        revng_assert(MaybeSize.has_value());
+
+        ReturnStruct->Size += MaybeSize.value();
       }
 
       revng_assert(ReturnStruct->Size != 0 && !ReturnStruct->Fields.empty());
 
-      if constexpr (!DryRun) {
-        auto ReturnStructTypePath = TheBinary.recordNewType(std::move(Result));
-        revng_assert(ReturnStructTypePath.isValid());
-        return model::QualifiedType{ ReturnStructTypePath, {} };
-      } else {
-        return model::QualifiedType{};
-      }
+      model::QualifiedType QualifiedResult(Binary.getTypePath(Result.get()),
+                                           {});
+      TypeDependencies.emplace_back(std::move(Result));
+      return QualifiedResult;
     }
 
     return std::nullopt;
   }
 
-  template<typename RType, size_t RCount>
-  static bool verifyArgumentsToBeConvertible(const SortedVector<RType> &UR,
-                                             const RegisterArray<RCount> &AR,
-                                             model::Binary &B) {
-    return convertArguments<RType, RCount, true>(UR, AR, B).has_value();
-  }
-
-  template<typename RType, size_t RCount>
-  static bool
-  verifyReturnValueToBeConvertible(const SortedVector<RType> &UR,
-                                   const RegisterArray<RCount> &AR,
-                                   const model::Register::Values PtC,
-                                   model::Binary &B) {
-    return convertReturnValue<RType, RCount, true>(UR, AR, PtC, B).has_value();
-  }
-
   static DistributedArguments
-  distributePositionBasedArguments(const ArgumentContainer &Arguments) {
+  distributePositionBasedArguments(const ArgumentVector &Arguments,
```

Comment:
Carrying out an open thread from #241: https://github.com/revng/revng/pull/241#discussion_r1009662370

Reply:
I'm pretty sure this is outside the scope of this PR after all and should be handled by one of the consequent ones.

@aleclearmind, please confirm/deny.

Reply:
Uhm, I think we can assume `xmm` is 128-bits and ignore any return value (assuming that's what we're talking about) that's not in a register we name in `Register.h`.

Reply:
The question was about whether we tackle in this PR or (my preferred option) pass it down further to the next one.

Reply:
What is the specific question? I'm a bit confused.

Reply:
The specific question is:

Is handling the stuff you clarified at the end of [the original thread](https://github.com/revng/revng/pull/241#discussion_r1011611897) and in [this message](https://github.com/revng/revng/pull/232#discussion_r1049870469) appropriate for this PR or should we just "ignore" them until the PR focused on floating point stuff?

Reply:
If it's a 3 lines change OK, if it encompasses complex reasoning about floats which we currently do not support very well, let's postpone.

Reply:
I'd imagine it's a couple of lines, yes. But there's currently no clear way to test all of it (the testing is still floating-point agnostic), so it's not verifiable whether those couple of lines are correct or not.

As such, I'd rather postpone doing anything of the sort until the testing is extended in the proper PR.

Also note that the current iteration of the analysis filters out any mentions of floating point stuff of not-fully-supported architectures before attempting to convert any functions.

Reply:
Then let's leave it out